### ISSUE TITLE
Load syntax schemes on demand for each file type

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#3493]
Right now, syntax highlighting loads all syntax schemes eagerly on first request:
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Highlighting/SyntaxHighlightingService.cs#L460

> VS bug #568061

**Comment by Therzok** [comment]
```
1.42%   TextEditorDisplayBinding..cctor  •  1,267 ms  •  1 call  •  MonoDevelop.Ide.Editor.TextEditorDisplayBinding..cctor
  1.42%   InitSourceEditor  •  1,267 ms  •  1 call  •  MonoDevelop.Ide.Editor.TextEditorDisplayBinding.InitSourceEditor
    1.42%   LoadCustomStylesAndModes  •  1,263 ms  •  1 call  •  MonoDevelop.Ide.Editor.TextEditorDisplayBinding.LoadCustomStylesAndModes
      1.42%   SyntaxHighlightingService..cctor  •  1,262 ms  •  1 call  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService..cctor
        1.20%   LoadStylesAndModes  •  1,065 ms  •  2 calls  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService.LoadStylesAndModes(Assembly)
          ►1.19%   LoadFile  •  1,060 ms  •  2,842 calls  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService.LoadFile(LanguageBundle, String, Func[Stream], Func[IStreamProvider])
          ►0.00%   SyntaxHighlightingService+<>c__DisplayClass31_1..ctor  •  1 ms  •  2,842 calls  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService+<>c__DisplayClass31_1..ctor
           0.00%   GetManifestResourceNames  •  0 ms  •  2 calls  •  System.Reflection.RuntimeAssembly.GetManifestResourceNames
           0.00%   CtorClosed  •  0 ms  •  5,684 calls  •  System.MulticastDelegate.CtorClosed(Object, IntPtr)
          ►0.00%   OnAssemblyLoadEvent  •  0 ms  •  1 call  •  System.AppDomain.OnAssemblyLoadEvent(RuntimeAssembly)
          ►0.00%   SyntaxHighlightingService+<>c__DisplayClass31_0..ctor  •  0 ms  •  2 calls  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService+<>c__DisplayClass31_0..ctor
        ►0.22%   PrepareMatches  •  195 ms  •  1 call  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService.PrepareMatches
        ►0.00%   AddExtensionNodeHandler  •  0 ms  •  1 call  •  Mono.Addins.AddinManager.AddExtensionNodeHandler(String, ExtensionNodeEventHandler)
        ►0.00%   FirstOrDefault  •  0 ms  •  1 call  •  System.Linq.Enumerable.FirstOrDefault(IEnumerable[TSource], Func[TSourceBoolean])
         0.00%   Regex..ctor  •  0 ms  •  3 calls  •  System.Text.RegularExpressions.Regex..ctor(String)
        ►0.00%   LanguageBundle..ctor  •  0 ms  •  1 call  •  MonoDevelop.Ide.Editor.Highlighting.LanguageBundle..ctor(String, String)
         0.00%   GetFiles  •  0 ms  •  1 call  •  System.IO.Directory.GetFiles(String)
         0.00%   Exists  •  0 ms  •  1 call  •  System.IO.Directory.Exists(String)
         0.00%   GetAssemblies  •  0 ms  •  1 call  •  System.AppDomain.GetAssemblies
        ►0.00%   get_LanguageBundlePath  •  0 ms  •  2 calls  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService.get_LanguageBundlePath
         0.00%   List`1..ctor  •  0 ms  •  1 call  •  System.Collections.Generic.List`1..ctor
         0.00%   CtorOpened  •  0 ms  •  1 call  •  System.MulticastDelegate.CtorOpened(Object, IntPtr, IntPtr)
         0.00%   Add  •  0 ms  •  1 call  •  System.Collections.Generic.List`1.Add(T)
         0.00%   get_CurrentDomain  •  0 ms  •  1 call  •  System.AppDomain.get_CurrentDomain
         0.00%   op_Inequality  •  0 ms  •  1 call  •  System.Reflection.Assembly.op_Inequality(Assembly, Assembly)
         0.00%   get_Assembly  •  0 ms  •  1 call  •  System.RuntimeType.get_Assembly
         0.00%   op_Implicit  •  0 ms  •  2 calls  •  MonoDevelop.Core.FilePath.op_Implicit(FilePath)
         0.00%   CtorClosed  •  0 ms  •  1 call  •  System.MulticastDelegate.CtorClosed(Object, IntPtr)
      ►0.00%   LoadStylesAndModesInPath  •  0 ms  •  1 call  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService.LoadStylesAndModesInPath(String)
       0.00%   Exists  •  0 ms  •  1 call  •  System.IO.Directory.Exists(String)
      ►0.00%   SyntaxHighlightingService+<>c..cctor  •  0 ms  •  1 call  •  MonoDevelop.Ide.Editor.Highlighting.SyntaxHighlightingService+<>c..cctor
      ►0.00%   get_SyntaxModePath  •  0 ms  •  2 calls  •  MonoDevelop.Ide.Editor.TextEditorDisplayBinding.get_SyntaxModePath
       0.00%   op_Implicit  •  0 ms  •  2 calls  •  MonoDevelop.Core.FilePath.op_Implicit(FilePath)
    ►0.00%   DefaultSourceEditorOptions..cctor  •  4 ms  •  1 call  •  MonoDevelop.Ide.Editor.DefaultSourceEditorOptions..cctor
     0.00%   Init  •  0 ms  •  1 call  •  MonoDevelop.Ide.Editor.DefaultSourceEditorOptions.Init
```

**Comment by mkrueger** [comment]
Measured the SyntaxHighlightingService constructor my system with stop watch : ~270ms

Consider that the profile tool is giving inaccurate results here - or my machine is really way more faster than yours. 

But yes iit could be optimized - maybe loading the shemes in parallel. Our old json format could be loaded lazily it was built for that - this not really possible with the textmate themes. So we need some sort of language theme registry here. But given the complexity of implementing that it's atm not worth it.

We'll switch to the vs.net textmate engine anytime soon so anything we would implement here now will be lost.

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.6.0.1573 (master)

Author: Mike Krüger
Commit: 547683df4a7ea7d436403020789e323c215cbb00 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.6.0.1755 (release-7.6)

Author: Mike Krüger
Commit: 547683df4a7ea7d436403020789e323c215cbb00 (mono/monodevelop)
